subsys/modem/modem_stats.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2024 Trackunit Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef _POSIX_VERSION
+#undef _POSIX_VERSION
+#endif
+#define _POSIX_VERSION 200809L
+
+#include <zephyr/modem/stats.h>
+#include <zephyr/shell/shell.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(modem_stats);
+
+static struct k_spinlock stats_buffer_lock;
+static sys_slist_t stats_buffer_list;
+
+static struct modem_stats_buffer *stats_buffer_from_node(sys_snode_t *node)
+{
+	return (struct modem_stats_buffer *)node;
+}
+
+static void stats_buffer_list_append(struct modem_stats_buffer *buffer)
+{
+	K_SPINLOCK(&stats_buffer_lock) {
+		sys_slist_append(&stats_buffer_list, &buffer->node);
+	}
+}
+
+static struct modem_stats_buffer *stats_buffer_list_first(void)
+{
+	struct modem_stats_buffer *first;
+
+	K_SPINLOCK(&stats_buffer_lock) {
+		first = stats_buffer_from_node(sys_slist_peek_head(&stats_buffer_list));
+	}
+
+	return first;
+}
+
+static struct modem_stats_buffer *stats_buffer_list_next(struct modem_stats_buffer *buffer)
+{
+	struct modem_stats_buffer *next;
+
+	K_SPINLOCK(&stats_buffer_lock) {
+		next = stats_buffer_from_node(sys_slist_peek_next(&buffer->node));
+	}
+
+	return next;
+}
+
+static uint8_t percent_used(uint32_t max, uint32_t cap)
+{
+	uint64_t percent;
+
+	if (max == 0) {
+		return 0;
+	}
+
+	if (max == cap) {
+		return 100;
+	}
+
+	percent = 100;
+	percent <<= 16;
+	percent /= cap;
+	percent *= max;
+	percent >>= 16;
```

Comment:
Simpler (with the multiplication first):
```suggestion
	percent = 100 * max / cap;
```

Reply:
I like to keep the types and operations really clear, as UINT32_MAX can't be multiplied by 100, without overflow, so it has to be stored in a `uint64_t`. It can be done as a 1-liner, but it looks like this
```
return (uint8_t)((((uint64_t)100) * max) / cap);
```
can we meet in the middle with this one
```
	percent = 100;
	percent *= max;
	percent /= cap;
```
The up/downshifting is a bit overkill anyway :)